### ISSUE TITLE
REST vs doing_front or admin (Gutenberg but not only)

### ISSUE THREAD

**Issue by herrvigg** [qtranslate/qtranslate-xt#744]
Among the complicated things coming from qTranslate-X, there is this concept of **doing_front**. The main goal is to differentiate the admin from standard requests by "normal" end users. The admin part includes WP_ADMIN, ~WP_AJAX, DOING_CRON~ and other very strange cases depending on the HTTP_REFERER that  to be honest i don't understand completely yet... 

During the init phase in the core module, after having gone through quite a complicated detection phase  on `plugins_loaded` (...which would require a ton of refactoring to make it readable by a normal human being), there is one moment where the loading sequence diverge between front and admin depending on that.

With the experiments for Gutenberg i came to a problematic question: should a **REST** call allow both doing_front and admin? This is not an easy question and it has quite some implications. In general, it should be ` doing_front`, which may sound surprising, but in reality the underlying requests come from a standard user. Though it's not clear if **doing_front** is really the right approach you want all the front hooks here or at least all the translation to be done as for a standard request.

*However* if we are dealing with the admin section, you might also do some REST calls from there. And obviously the URL does not tell you anything about this since you are on `your-site.com/wp-json/whatever-endpoint`.

This is the case with Gutenberg making tons of REST calls from the admin section. And in that case you would rather want to enter the admin branch. But it's not clear at all how to deal with that! Should it come from:
- explicit query argument from GET or POST
- HTTP_REFERER
- some hook mechanism... but how to make it RESTFUL?
- ~cookies~ nooo! In general cookies are not RESTFUL, if they are keeping a server state. The only exception should be for authentication or just a fallback when it's a client state. This is a specific topic that should be addressed btw.

**Comment by herrvigg** [comment]
It seems that WordPress updates the HTTP_REFERER correctly on that case but qTranslate does not handle it properly. Example of a REST call from Gutenberg (admin):
```
Request URL: http://localhost:8000/wp-json/wp/v2/posts/24056?_locale=user
Referer: http://localhost:8000/wp-admin/post.php?post=24056&action=edit&lang=en
```

In that case i think qTranslate should rather switch to the admin part, which for some reason it apparently does not do currently... This is probably the easiest way, as in most of the cases you don't have the possibility to patch the request URL if it comes from WordPress code. So it would actually be a bug in qTranslate. Do you agree with that?

**Comment by herrvigg** [comment]
This problem is actually a sub-case of #720 when the language is not defined. I think HTTP_REFERER is the good solution, but testing all the different cases is **extremely** complicated without proper unit tests.

**Comment by herrvigg** [comment]
Released in 3.7.1.
